crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -117,46 +144,75 @@ impl GatewayQuery {
             });
         }
 
-        // Package names that we have or will issue requests for.
-        let mut seen = HashSet::new();
-        let mut pending_package_specs = HashMap::new();
-        for spec in self.specs {
-            if let Some(name) = &spec.name {
-                seen.insert(name.clone());
-                pending_package_specs
-                    .entry(name.clone())
-                    .or_insert_with(Vec::new)
-                    .push(spec);
-            }
-        }
-
-        // A list of futures to fetch the records for the pending package names. The main task
-        // awaits these futures.
+        // A list of futures to fetch the records for the pending package names.
+        // The main task awaits these futures.
         let mut pending_records = FuturesUnordered::new();
 
-        // The resulting list of repodata records.
-        let mut result = vec![RepoData::default(); subdirs.len()];
+        // Push the direct url queries to the pending_records.
+        for spec in direct_url_specs.clone() {
+            let gateway = self.gateway.clone();
+            pending_records.push(
+                async move {
+                    let url = spec.1;
+                    let query = DirectUrlQuery::new(
+                        url.clone(),
+                        gateway.package_cache.clone(),
+                        gateway.client.clone(),
+                    );
+
+                    let record = query
+                        .execute()
+                        .await
+                        .map_err(|e| GatewayError::DirectUrlQueryError(url.to_string(), e))?;
+
+                    // Check if record actually has the same name
+                    if let Some(record) = record.first() {
+                        let spec_name = spec
+                            .0
+                            .clone()
+                            .name
```

Comment:
Here you also extract this information for a second time. Perhaps also store in `direct_url_specs`?

Reply:
https://github.com/mamba-org/rattler/pull/725/commits/7ea414a9b3ecf609d1c7027ad99a9e99c50e232